src/skins/vector/views/organisms/UserSettings.js

```diff
@@ -15,15 +15,144 @@ limitations under the License.
     displayName: 'UserSettings',
-    mixins: [UserSettingsController],
+
+    Phases: {
+        Loading: "loading",
+        Saving: "saving",
+        Display: "display",
+    },
+
+    getInitialState: function() {
+        return {
+            avatarUrl: null,
+            displayName: null,
+            threePids: [],
+            clientVersion: version,
+            phase: this.Phases.Loading,
+        };
+    },
+
+    componentWillMount: function() {
+        var self = this;
+
+        var profilePromise = UserSettingsStore.loadProfileInfo();
+        var threepidPromise = UserSettingsStore.loadThreePids();
+
+        q.all([profilePromise, threepidPromise]).then(
+            function(resps) {
+                self.setState({
+                    avatarUrl: resps[0].avatar_url,
+                    displayName: resps[0].displayname,
+                    threepids: resps[1].threepids,
+                    phase: self.Phases.Display,
+                });
+
+                // keep a copy of the original state in order to track changes
+                self.setState({
+                    originalState: self.state
+                });
+            },
+            function(error) {
+                var ErrorDialog = sdk.getComponent("organisms.ErrorDialog");
+                Modal.createDialog(ErrorDialog, {
+                    title: "Can't load user settings",
+                    description: error.toString()
+                });
+            }
+        );
+    },
+
+    componentDidMount: function() {
+        this.dispatcherRef = dis.register(this.onAction);
+    },
+
+    componentWillUnmount: function() {
+        dis.unregister(this.dispatcherRef);
+    },
+
+    onSaveClicked: function(ev) {
+        var self = this;
+        var savePromises = [];
+
+        // XXX: this is managed in ChangeAvatar.js, although could be moved out here in order
+        // to allow for the change to be staged alongside the rest of the form.
+        //
+        // if (this.state.originalState.avatarUrl !== this.state.avatarUrl) {
+        //     savePromises.push( UserSettingsStore.saveAvatarUrl(this.state.avatarUrl) );
+        // }
+
+        if (this.state.originalState.displayName !== this.state.displayName) {
+            savePromises.push( UserSettingsStore.saveDisplayName(this.state.displayName) );
+        }
+
+        if (this.state.originalState.threepids.length !== this.state.threepids.length ||
+            this.state.originalState.threepids.every(function(element, index) {
+                    return element === this.state.threepids[index];
+            }))
+        {
+            savePromises.push( UserSettingsStore.saveThreePids(this.state.threepids) );
+        }
+
+        // TODO: do the password check
+
+        self.setState({
+            phase: self.Phases.Saving,
+        });
+
+        q.all(savePromises).then(
+            function(resps) {
+                self.setState({
+                    phase: self.Phases.Display,
```

Comment:
We're keeping a state machine around in this component so we know what to display. This state machine may be more useful if it was moved to the `UserSettingsStore` coupled with emitted actions. This would keep this component "dumb" as it would just be updating the display depending on what came out from the SDK. Something like:
```javascript
getInitialState() {
  return { userSettings: null, currentAction: LOADING_USER_SETTINGS }
}

componentWillMount() {
  UserSettingsStore.getUserSettings().done(function(userSettings) {
    this.setState({userSettings: userSettings});
  });
}

onAction(payload) {
  this.setState({currentAction: payload.action});
}

render() {
  switch (currentAction) {
    case LOADING_USER_SETTINGS:
      // display global spinner
    case SAVING_USER_SETTINGS:
      // display saving spinner
    case USER_SETTINGS:
      // display user settings
  }
}
```

YMMV - It may be worth keeping the state machine around if we're doing more complicated things, but it feels like this is a pretty common operation that the React SDK should probably be doing it rather than a specific component.